File: python/examples/completion.py
"""Shape completion and per-point uncertainty from a partial observation.

Synthesizes a shape from a small linear model, hides part of it, registers
the atlas to the partial cloud, then completes the missing region and reads
off a per-point confidence map. Run: python examples/completion.py
"""

import numpy as np

import rustcpd as cpd


def model(m, k):
    z = (np.arange(m) + 1.0)[:, None]
    mean = np.column_stack(
        (np.sin(z[:, 0] * 0.3) * 1.5, np.cos(z[:, 0] * 0.2), np.sin(z[:, 0] * 0.13))
    )
    rows = np.arange(m * 3)[:, None]
    modes = np.sin((rows + 1 + 5 * np.arange(k)) * 0.17) + 0.3 * np.cos(
        (rows + 2 * np.arange(k) + 1) * 0.09
    )
    modes, _ = np.linalg.qr(modes)  # orthonormal columns
    eigenvalues = np.array([0.6 / (c + 1) for c in range(k)])
    return mean, modes, eigenvalues


def main():
    m, k = 80, 5
    mean, modes, eigenvalues = model(m, k)
    truth = np.array([0.4, -0.3, 0.2, -0.12, 0.08])

    # A true shape, posed into the target frame.
    deformed = mean + (modes @ truth).reshape(m, 3)
    angle = 0.25
    r = np.array(
        [[np.cos(angle), -np.sin(angle), 0.0],
         [np.sin(angle), np.cos(angle), 0.0], [0.0, 0.0, 1.0]]
    )
    full = 1.15 * (deformed @ r) + np.array([0.3, -0.2, 0.1])

    # Observe only the first 55 of 80 points (a partial scan).
    observed = 55
    partial = full[:observed]

    fit = cpd.register_atlas(
        partial, mean, modes, eigenvalues,
        lambda_regularization=1.0, optimize_similarity=True,
        tolerance=1e-8, max_iterations=200,
    )

    # Complete the shape and get the confidence map.
    post = fit.posterior(partial, mean, modes, eigenvalues, completeness=observed / m)
    completed = post.predict()               # (M, 3), target frame
    variance = post.predictive_variance()    # (M,), per-point total variance

    held = slice(observed, m)
    completion_rms = np.sqrt(np.mean((completed[held] - full[held]) ** 2))
    posed_mean = fit.scale * (mean @ fit.rotation) + fit.translation
    baseline_rms = np.sqrt(np.mean((posed_mean[held] - full[held]) ** 2))
    print(f"held-out completion rms = {completion_rms:.3e} "
          f"(posed-mean baseline {baseline_rms:.3e})")
    print(f"mean predictive sd: observed {np.sqrt(variance[:observed].mean()):.3e}, "
          f"missing {np.sqrt(variance[observed:].mean()):.3e}")

    # An ensemble of plausible completions (for bands / downstream stats).
    ensemble = post.sample_shapes(200, seed=0)
    spread = np.std([s[held] for s in ensemble], axis=0).mean()
    print(f"ensemble of {len(ensemble)} completions; mean held-out spread {spread:.3e}")


if __name__ == "__main__":
    main()

File: rustcpd/examples/completion.rs
//! Shape completion + per-point uncertainty from a partial observation.
//!
//! Run: `cargo run --release --features completion --example completion`

use rustcpd::{AtlasConfig, AtlasRegistration, DMatrix, EmConfig, PosteriorOptions};

fn model(m: usize, k: usize) -> (DMatrix<f64>, DMatrix<f64>, Vec<f64>) {
    let mean = DMatrix::from_fn(m, 3, |i, j| {
        let z = (i + 1) as f64;
        match j {
            0 => (z * 0.3).sin() * 1.5,
            1 => (z * 0.2).cos(),
            _ => (z * 0.13).sin(),
        }
    });
    let mut modes = DMatrix::from_fn(m * 3, k, |i, c| {
        ((i + 1 + c * 5) as f64 * 0.17).sin() + 0.3 * ((i + 2 * c + 1) as f64 * 0.09).cos()
    });
    for c in 0..k {
        for prev in 0..c {
            let proj = (0..m * 3)
                .map(|i| modes[(i, c)] * modes[(i, prev)])
                .sum::<f64>();
            for i in 0..m * 3 {
                modes[(i, c)] -= proj * modes[(i, prev)];
            }
        }
        let norm = (0..m * 3)
            .map(|i| modes[(i, c)].powi(2))
            .sum::<f64>()
            .sqrt();
        for i in 0..m * 3 {
            modes[(i, c)] /= norm;
        }
    }
    let eigenvalues = (0..k).map(|c| 0.6 / (c + 1) as f64).collect();
    (mean, modes, eigenvalues)
}

fn main() -> Result<(), rustcpd::Error> {
    let (m, k) = (80, 5);
    let (mean, modes, eigenvalues) = model(m, k);
    let truth = [0.4, -0.3, 0.2, -0.12, 0.08];

    // True shape, posed into the target frame.
    let deformed = DMatrix::from_fn(m, 3, |i, j| {
        mean[(i, j)]
            + (0..k)
                .map(|c| modes[(i * 3 + j, c)] * truth[c])
                .sum::<f64>()
    });
    let angle = 0.25_f64;
    let r = DMatrix::from_row_slice(
        3,
        3,
        &[
            angle.cos(),
            -angle.sin(),
            0.0,
            angle.sin(),
            angle.cos(),
            0.0,
            0.0,
            0.0,
            1.0,
        ],
    );
    let full = DMatrix::from_fn(m, 3, |i, j| {
        1.15 * (0..3).map(|q| deformed[(i, q)] * r[(q, j)]).sum::<f64>() + [0.3, -0.2, 0.1][j]
    });

    // Observe only the first 55 of 80 points.
    let observed = 55;
    let partial = DMatrix::from_fn(observed, 3, |i, j| full[(i, j)]);

    let fit = AtlasRegistration::new(
        &partial,
        &mean,
        &modes,
        AtlasConfig {
            em: EmConfig {
                tolerance: 1e-8,
                max_iterations: 200,
                ..Default::default()
            },
            eigenvalues: eigenvalues.clone(),
            lambda_regularization: 1.0,
            optimize_similarity: true,
            ..Default::default()
        },
    )?
    .register()?;

    let posterior = fit.posterior(
        &partial,
        &mean,
        &modes,
        &eigenvalues,
        &PosteriorOptions {
            completeness: Some(observed as f64 / m as f64),
            ..Default::default()
        },
    )?;

    let completed = posterior.predict();
    let held_rms = ((observed..m)
        .map(|i| {
            (0..3)
                .map(|j| (completed[(i, j)] - full[(i, j)]).powi(2))
                .sum::<f64>()
        })
        .sum::<f64>()
        / (3 * (m - observed)) as f64)
        .sqrt();
    println!("held-out completion rms = {held_rms:.3e}");

    let variance = posterior.predictive_variance();
    let obs_sd = (variance[..observed].iter().sum::<f64>() / observed as f64).sqrt();
    let miss_sd = (variance[observed..].iter().sum::<f64>() / (m - observed) as f64).sqrt();
    println!("mean predictive sd: observed {obs_sd:.3e}, missing {miss_sd:.3e}");

    // Ensemble of plausible completions (deterministic in the seed).
    let ensemble = posterior.sample_shapes(200, 0)?;
    println!("drew {} plausible completions", ensemble.len());
    Ok(())
}
